docs/config-luauconfig.md

```diff
@@ -0,0 +1,129 @@
+
+## Design
+
+We introduce a new configuration file called `.luauconfig`.
+
+### Syntax
+
+An example `.luauconfig` is given below.
+
+```luau
+return {
+    languageMode = "nonstrict",
+    lint = {
+        ["*"] = true,
+        LocalUnused = false
+    },
+    lintErrors = true,
+    typeErrors = true,
+    globals = {"expect"},
+    aliases = {
+        src = "./src"
+    }
+}
+```
+
+The file must return a Luau table of the following `LuauConfig` type:
+```luau
+type LanguageMode = "strict" | "nonstrict" | "nocheck"
+
+type LintWarning =
+    "*"
+    | "UnknownGlobal"
+    | "DeprecatedGlobal"
+    | "GlobalUsedAsLocal"
+    | "LocalShadow"
+    | "SameLineStatement"
+    | "MultiLineStatement"
+    | "LocalUnused"
+    | "FunctionUnused"
+    | "ImportUnused"
+    | "BuiltinGlobalWrite"
+    | "PlaceholderRead"
+    | "UnreachableCode"
+    | "UnknownType"
+    | "ForRange"
+    | "UnbalancedAssignment"
+    | "ImplicitReturn"
+    | "DuplicateLocal"
+    | "FormatString"
+    | "TableLiteral"
+    | "UninitializedLocal"
+    | "DuplicateFunction"
+    | "DeprecatedApi"
+    | "TableOperations"
+    | "DuplicateCondition"
+    | "MisleadingAndOr"
+    | "CommentDirective"
+    | "IntegerParsing"
+    | "ComparisonPrecedence"
+    | "RedundantNativeAttribute"
+
+type LuauConfig = {
+    languageMode: LanguageMode?,
+    lint: { [LintWarning]: boolean }?,
+    lintErrors: boolean?,
+    typeErrors: boolean?,
+    globals: { string }?,
+    aliases: { [string]: string }?,
+}
+```
+
+If configuration fields are not provided, `.luauconfig` uses the same default behavior as `.luaurc`.
+
+### Search semantics
+
+The search semantics of `.luauconfig` files are the same as `.luaurc`'s:
+
+> For a given .lua file, Luau will search for .luaurc files starting from the folder that the .lua file is in; all files in the ancestry chain will be parsed and their configuration applied.
+> When multiple files are used, the file closer to the .lua file overrides the settings.
+
+However, if both `.luaurc` and `.luauconfig` files are present in a directory, only the `.luauconfig` file is used.
+Configuration files are not merged; at most one configuration file is recognized per directory.
+
+### Configuration extraction
+
+Unlike `.luaurc` files, `.luauconfig` files can contain runtime constructs like variables, functions, and loops.
+To extract configuration from a `.luauconfig` file, its contents are executed in an isolated Luau VM with only the standard libraries enabled (`luaL_openlibs`).
+This approach ensures that configuration files can leverage Luau features while maintaining a sandboxed execution environment.
+
+As with the [user-defined type functions RFC](https://rfcs.luau.org/user-defined-type-functions.html), a time limit is imposed on the execution of `.luauconfig` files to guard against excessive computation.
```

Comment:
User-defined type functions do not have a timeout. They will run indefinitely unless the embedding environment asks them to stop. This works for analysis-time-only things, but the config affects runtime semantics as well.